### PR TITLE
Fixed some issues of santi's PR

### PR DESCRIPTION
**Pull request by coolo**

### CODE REVIEW

File: etc/openqa/workers.ini [hunk 1]
```diff
@@ -23,3 +23,8 @@
 
 #[openqa.example.host]
 # SHARE_DIRECTORY = /var/lib/openqa/example
+# In case caching of assets and tests needs to be enabled
+# CACHEDIRECTORY and TESTPOOLSERVER need to be enabled
+# CACHEDIRECTORY = /var/lib/openqa/cache
+# TESTPOOLSERVER = rsync://deimos.suse.de/openqa-tests
```
Comment (okurz) [inline etc/openqa/workers.ini]: can this be described in more details what it is supposed to mean?

File: lib/OpenQA/Cache.pm [hunk 2]
```diff
@@ -0,0 +1,197 @@
+package OpenQA::Cache;
```
Comment (okurz) [inline lib/OpenQA/Cache.pm]: copyright header please

File: lib/OpenQA/Cache.pm [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        else {
+            $type = "Cache Miss";
+            download_asset($job->{id}, lc($asset_type), $asset);
+            set($asset);
+        }
+
+        write_db();
+        close($asset_fd);
+        last;
+
```
Comment (okurz) [inline lib/OpenQA/Cache.pm]: can you delete some blank lines please?

File: lib/OpenQA/Cache.pm [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        log_debug("Read cache db file");
+    }
+    else {
+        log_debug "Creating empty cache";
+        write_db;
+    }
+
+
+
+}
```
Comment (okurz) [inline lib/OpenQA/Cache.pm]: remove blank lines please

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 5]
```diff
@@ -62,9 +65,47 @@ sub _save_vars($) {
...
+    return undef;
+}
+
+sub cache_tests {
+
+    my ($shared_cache, $testpoolserver) = @_;
+
+    my $start = time;
+    #Do an flock to ensure only one worker is trying to synchronize at a time.
+    my @cmd = qw(flock -E 999);
```
Comment (okurz) [inline lib/OpenQA/Worker/Engines/isotovideo.pm]: so that would serialize the syncing job of all workers and hinder scalability, right? Why is this necessary? The worker should access the server r/o anyway so for reasons to serialize the I/O on the server?

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 6]
```diff
@@ -62,9 +65,47 @@ sub _save_vars($) {
...
+    push @cmd, "$shared_cache/needleslock";
+    push @cmd, (qw(rsync -aP), $testpoolserver, qw(--delete --exclude=.git));
```
Comment (okurz) [inline lib/OpenQA/Worker/Engines/isotovideo.pm]: I suggest `-aHP` to preserve hard-links

File: lib/OpenQA/Worker/Engines/isotovideo.pm [hunk 7]
```diff
@@ -146,7 +172,7 @@ sub engine_workit($) {
         setpgrp(0, 0);
         $ENV{TMPDIR} = $tmpdir;
         log_info("$$: WORKING " . $job->{id});
-        if (open(my $log, '>', "autoinst-log.txt")) {
+        if (open(my $log, '>>', "autoinst-log.txt")) {
             print $log "+++ worker notes +++\n";
```
Comment (okurz) [inline lib/OpenQA/Worker/Engines/isotovideo.pm]: so we have +++ setup notes… and then later one +++ worker notes with partially duplicated content. Can this be improved?

File: t/full-stack.t [hunk 8]
```diff
@@ -279,6 +278,27 @@ close($f);
...
+[http://localhost:$mojoport]
+TESTPOOLSERVER = $cwd/t/full-stack.d/openqa/share/tests
+EOC
+close($conf);
+
+start_worker;
+client_call('jobs/3/restart post', qr{\Qtest_url => ["/tests/5\E}, 'client returned new test_url');
+
+print STDERR "SLEEP\n";
+sleep 100;
```
Comment (okurz) [inline t/full-stack.t]: ???

File: etc/openqa/workers.ini (hunk 1 above)
Comment (coolo) [inline etc/openqa/workers.ini]: http://changingminds.org/techniques/questioning/open_closed_questions.htm

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 5 above)
Comment (coolo) [inline lib/OpenQA/Worker/Engines/isotovideo.pm]: This is not hindering scalability - there is only one directory to sync and rsync will abort if you sync it twice in parallel.

File: lib/OpenQA/Worker/Engines/isotovideo.pm (hunk 7 above)
Comment (coolo) [inline lib/OpenQA/Worker/Engines/isotovideo.pm]: possibly later

File: lib/OpenQA/Cache.pm (hunk 2 above)
Comment (coolo) [inline lib/OpenQA/Cache.pm]: Looks like Santi started from Utils.pm - which missed it too :)

**Comment by coolo** [comment]
As you might see on the number of commits, there were some more issues :)

And the cache implementation doesn't look proper to me, but I'm struggling on how to test this properly.

**Comment by okurz** [comment]
from my POV there are either less issues observable on osd right now which can be regarded to this or people just calmed down a bit :-) So I don't see a need for rushing and rather would like to not piss off reviewers by new non-understood errors. So better enjoy the weekend differently ;-)

**Comment by coolo** [comment]
you're implying I enjoy my weekends most if I piss off reviewers?

**Comment by okurz** [comment]
not at all. That would be your work week ;-)

No. I meant by unforeseen errors due to only you testing over the weekend there might be some issues missed.

**Comment by coolo** [comment]
one remaining issue is that the rsync call blocks us to death

**Comment by coolo** [comment]
Another rather unexpected issue:
```
QEMU: qemu-system-x86_64: -drive media=cdrom,if=none,id=cd0,format=raw,file=SLE12-SP3-Staging:B-Test-Server-DVD-x86_64-Build14.13-Media.iso: Unknown protocol 'SLE12-SP3-Staging'

File: etc/openqa/workers.ini (hunk 1 above)
Comment (okurz) [inline etc/openqa/workers.ini]: ok, let me retry: I am sorry, I don't understand what `TESTPOOLSERVER` means just by reading the name and I am saying this because I think others will have the same problem. I would appreciate if `TESTPOOLSERVER` is described in more details.

@coolo What would be your approach how to phrase a request like the above into an open question which makes the requestee think but also incorporate the suggestions?

Comment (coolo) [inline etc/openqa/workers.ini]: My approach would sound like:

I think this is hard to understand without context. Can you add more descriptions what TESTPOOLSERVER stands for?

Comment (okurz) [inline etc/openqa/workers.ini]: but that is a closed question, I don't understand

Comment (coolo) [inline etc/openqa/workers.ini]: I didn't say asking open questions is easy :)

Comment (coolo) [inline etc/openqa/workers.ini]: https://github.com/coolo/openQA/blob/3a8c12e23652ee23ea798faf791c3ac91e3ba4ca/etc/openqa/workers.ini - TESTPOOLSERVER is a crappy name. If someone has a better name, shoot. But we wanted to avoid rsync in the name to have the option to extend this later to e.g. also use http.

**Comment by coolo** [comment]
One thing I forgot:
```
cksum: /var/lib/openqa/share/factory/hdd/SLES-12-SP1-x86_64.qcow2.TEMP-wumwcCel: No such file or directory
```

The webui returns the *local* path of the tmp asset and the worker is running chksum on it - blindly assuming he can access it. I.e. this means it was already broken with multiple webuis.

**Comment by kravciak** [comment]
Could this cause missing assets/other uploaded by parent job?
https://openqa.suse.de/tests/833920#step/installation/13

**Comment by coolo** [comment]
Yes, see https://progress.opensuse.org/issues/17812 (linked from above test :)
